packages/typespec-ts/src/modular/emitModels.ts

```diff
@@ -1,10 +1,123 @@
 
+// ====== UTILITIES ======
+
+function isAzureCoreErrorSdkType(t: Type) {
+  return (
+    t.name &&
+    ["error", "errormodel", "innererror", "errorresponse"].includes(
+      t.name.toLowerCase()
+    ) &&
+    t.isCoreErrorType === true
+  );
+}
+
+function getCoreClientErrorType(name: string, coreClientTypes: Set<string>) {
+  const coreClientType: string = name === "Error" ? "ErrorModel" : name;
+  coreClientTypes.add(coreClientType);
+  return coreClientType;
+}
+
+// ====== TYPE EXTRACTION ======
+
+function extractModels(codeModel: ModularCodeModel): Type[] {
+  const models = codeModel.types.filter(
+    (t) =>
+      (t.type === "model" || t.type === "enum") && !isAzureCoreErrorSdkType(t)
+  );
+
+  for (const model of codeModel.types) {
+    if (model.type === "combined" && model.nullable) {
+      for (const unionModel of model.types ?? []) {
+        if (unionModel.type === "model") {
+          models.push(unionModel);
+        }
+      }
+    }
+  }
+  return models;
+}
+
+// ====== TYPE BUILDERS ======
+function buildEnumModel(
+  model: Type
+): OptionalKind<TypeAliasDeclarationStructure> {
+  return {
+    name: model.name!,
+    isExported: true,
+    docs: [
+      ...getDocsFromDescription(model.description),
+      // If it is a fixed enum we don't need to list the known values in the docs as the
+      // output will be a literal union which is self documenting
+      model.isFixed
+        ? ""
+        : // When we generate an "extensible" enum, the type will be "string" so we list the known values
+          // in the docs for user reference.
+          (model.values ?? []).map((v) => `"${v.value}"`).join(", ")
+    ],
+    type: model.isFixed
+      ? (model.values ?? []).map((v) => `"${v.value}"`).join(" | ")
+      : "string"
+  };
+}
+
+type InterfaceStructure = OptionalKind<InterfaceDeclarationStructure> & {
+  extends: string[];
+};
+
+function buildModelInterface(
+  model: Type,
+  cache: { coreClientTypes: Set<string> }
+): InterfaceStructure {
+  const modelProperties = model.properties ?? [];
+  const modelInterface = {
+    name: model.name ?? "FIXMYNAME",
+    isExported: true,
+    docs: getDocsFromDescription(model.description),
+    extends: [] as string[],
+    properties: (modelProperties ?? []).map((p) => {
+      const propertyMetadata = getType(p.type, p.format);
+      let propertyTypeName = propertyMetadata.name;
+      if (isAzureCoreErrorSdkType(p.type)) {
+        propertyTypeName = isAzureCoreErrorSdkType(p.type)
+          ? getCoreClientErrorType(propertyTypeName, cache.coreClientTypes)
+          : propertyTypeName;
+      }
+      if (propertyMetadata.nullable) {
+        propertyTypeName = `(${propertyTypeName} | null)`;
+      }
+      if (propertyMetadata.modifier === "Array") {
+        propertyTypeName = `${propertyTypeName}[]`;
+      }
+
+      if (propertyMetadata.modifier === "NullableArray") {
+        propertyTypeName = `(${propertyTypeName} | null)[]`;
+      }
```

Comment:
I wonder if we could use getTypeName here ?

Reply:
See my comment above, I removed getTypeName but will try to bring it back